File: src/notion_extractor/notion/fetch.py
import asyncio
from typing import Any

from notion_extractor.notion.client import AsyncNotionClient


async def fetch_block_children(
    client: AsyncNotionClient, block_id: str
) -> list[dict[str, Any]]:
    blocks = await _paginate_get(client, f"/blocks/{block_id}/children")

    async def hydrate(block: dict[str, Any]) -> dict[str, Any]:
        block.setdefault("children", [])
        if block.get("has_children") and block.get("type") not in {
            "child_page",
            "child_database",
        }:
            block["children"] = await fetch_block_children(client, block["id"])
        return block

    return list(await asyncio.gather(*(hydrate(b) for b in blocks)))


async def query_database(
    client: AsyncNotionClient, database_id: str
) -> list[dict[str, Any]]:
    return await _paginate_post(client, f"/databases/{database_id}/query", body={})


async def get_page(client: AsyncNotionClient, page_id: str) -> dict[str, Any]:
    return await client.get(f"/pages/{page_id}")


async def get_database(client: AsyncNotionClient, database_id: str) -> dict[str, Any]:
    return await client.get(f"/databases/{database_id}")


async def _paginate_get(
    client: AsyncNotionClient, path: str
) -> list[dict[str, Any]]:
    results: list[dict[str, Any]] = []
    cursor: str | None = None
    while True:
        qs = "?page_size=100" + (f"&start_cursor={cursor}" if cursor else "")
        page = await client.get(path + qs)
        results.extend(page.get("results", []))
        if not page.get("has_more"):
            return results
        cursor = page.get("next_cursor")


async def _paginate_post(
    client: AsyncNotionClient, path: str, *, body: dict[str, Any]
) -> list[dict[str, Any]]:
    results: list[dict[str, Any]] = []
    cursor: str | None = None
    while True:
        page_body: dict[str, Any] = {**body, "page_size": 100}
        if cursor:
            page_body["start_cursor"] = cursor
        page = await client.post(path, page_body)
        results.extend(page.get("results", []))
        if not page.get("has_more"):
            return results
        cursor = page.get("next_cursor")

File: app/src/server/notion/fetch.ts
import type { NotionClient } from './client.js';

// biome-ignore lint/suspicious/noExplicitAny: Notion blocks are dynamic JSON
type Block = Record<string, any>;

export async function fetchBlockChildren(client: NotionClient, blockId: string): Promise<Block[]> {
  const blocks = await paginateGet(client, `/blocks/${blockId}/children`);
  await Promise.all(
    blocks.map(async (b: Block) => {
      b.children ??= [];
      if (b.has_children && b.type !== 'child_page' && b.type !== 'child_database') {
        b.children = await fetchBlockChildren(client, b.id);
      }
    }),
  );
  return blocks;
}

export async function queryDatabase(client: NotionClient, databaseId: string): Promise<Block[]> {
  return paginatePost(client, `/databases/${databaseId}/query`, {});
}

export async function getPage(client: NotionClient, pageId: string): Promise<Block> {
  return client.get(`/pages/${pageId}`);
}

export async function getDatabase(client: NotionClient, databaseId: string): Promise<Block> {
  return client.get(`/databases/${databaseId}`);
}

interface NotionPage {
  results: Block[];
  has_more: boolean;
  next_cursor: string | null;
}

async function paginateGet(client: NotionClient, basePath: string): Promise<Block[]> {
  const results: Block[] = [];
  let cursor: string | null = null;
  for (;;) {
    const qs: string = `?page_size=100${cursor ? `&start_cursor=${cursor}` : ''}`;
    const page: NotionPage = await client.get<NotionPage>(basePath + qs);
    results.push(...(page.results ?? []));
    if (!page.has_more) return results;
    cursor = page.next_cursor;
  }
}

async function paginatePost(
  client: NotionClient,
  reqPath: string,
  body: Record<string, unknown>,
): Promise<Block[]> {
  const results: Block[] = [];
  let cursor: string | null = null;
  for (;;) {
    const pageBody: Record<string, unknown> = { ...body, page_size: 100 };
    if (cursor) pageBody.start_cursor = cursor;
    const page: NotionPage = await client.post<NotionPage>(reqPath, pageBody);
    results.push(...(page.results ?? []));
    if (!page.has_more) return results;
    cursor = page.next_cursor;
  }
}
